### PR TITLE
dts: Add new DTS/binding parser

### PR DESCRIPTION
**Pull request by ulfalizer**
```
dts: Add new DTS/binding parser

Add a new DTS/binding parser to scripts/dts/ for generating
generated_dts_board.conf and generated_dts_board_unfixed.h.

The old code is kept to generate some deprecated defines, using the
--deprecated-only flag. It will be removed later.

The new parser is implemented in three files in scripts/dts/:

dtlib.py:
  A low-level .dts parsing library. This is similar to devicetree.py in
  the old code, but is a general robust DTS parser that doesn't rely on
  preprocessing.

edtlib.py (e for extended):
  A library built on top of dtlib.py that brings together data from DTS
  files and bindings and creates Device instances with all the data for
  a device.

gen_defines.py:
  A script that uses edtlib.py to generate generated_dts_board.conf and
  generated_dts_board_unfixed.h. Corresponds to extract_dts_includes.py
  and the files in extract/ in the old code.

  Note that dtlib.py and edtlib.py are general libraries (similar to
  Kconfiglib) that could be used for other stuff besides gen_defines.py.

testdtlib.py:
  Test suite for dtlib.py. Can be run directly as a script.

testedtlib.py (uses test.dts and test-bindings/):
  Test suite for edtlib.py. Can be run directly as a script.

The test suites will be run automatically in CI.

The new code turns some things that were warnings (or not checked) in
the old code into errors, like missing properties specified with
'category: required' in the binding for the node.

The code includes lots of documentation and tries to give helpful error
messages instead of Python errors.

Co-authored-by: Kumar Gala <kumar.gala@linaro.org>
Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>
```

### CODE REVIEW

File: scripts/dts/extract_dts_includes.py [hunk 1]
```diff
@@ -533,7 +533,7 @@ def main():
 
     if args.include is not None:
         with open(args.include, 'w', encoding='utf-8') as f:
-            write_header(f, args.deprecate_only)
+            write_header(f, args.deprecated_only)
```
Comment (galak) [inline scripts/dts/extract_dts_includes.py]: Pull this out into its own commit/PR.

File: cmake/dts.cmake [hunk 2]
```diff
@@ -175,6 +198,14 @@ if(SUPPORTS_DTS)
     message(FATAL_ERROR "command failed with return code: ${ret}")
   endif()
 
+  #
+  # Add the legacy identifiers
+  #
+
+  file(APPEND ${GENERATED_DTS_BOARD_UNFIXED_H}
```
Comment (galak) [inline cmake/dts.cmake]: what is this for?

**Review by galak — CHANGES_REQUESTED** [review]
a few comments

File: cmake/dts.cmake (hunk 2 above)
Comment (ulfalizer) [reply cmake/dts.cmake]: Wrote about it on Slack. Seems the bsim tests (part of CI) do something funky where `generated_dts_board.h.deprecated` never gets generated. You get an #include error for it with the old approach.

File: scripts/dts/extract_dts_includes.py (hunk 1 above)
Comment (ulfalizer) [reply scripts/dts/extract_dts_includes.py]: Done

File: cmake/dts.cmake (hunk 2 above)
Comment (galak) [reply cmake/dts.cmake]: Need to understand this better, I'm not in favor of this solution, and unclear why this makes 'bsim' happy.

Comment (ulfalizer) [reply cmake/dts.cmake]: @galak 
Are you looking into it?

The header doesn't show up, so it breaks otherwise.

**Review by galak — APPROVED** [review]

**Review by pabigot — APPROVED** [review]
Review based on execution test only (didn't even look at the code).  

This worked nicely, including diagnosing a missing property, which cleans up one problem.

But it's missing the other half: it doesn't diagnose presence of property that is not in the yaml, such as a misspelled optional property.

That lack isn't worth blocking this, but if necessary I'll create an issue for it.

**Comment by ulfalizer** [comment]
> But it's missing the other half: it doesn't diagnose presence of property that is not in the yaml, such as a misspelled optional property.
> 
> That lack isn't worth blocking this, but if necessary I'll create an issue for it.

Would be easy to add later, if we want to require all properties to be listed in the binding. Suspect we might have lots of nodes that don't follow that now.

**Comment by pabigot** [comment]
> Would be easy to add later, if we want to require all properties to be listed in the binding. Suspect we might have lots of nodes that don't follow that now.

Probably not, unless this implementation (unlike the current one) generates defines for the extra properties.  If there are any, they can't be being used.

**Comment by ulfalizer** [comment]
> > Would be easy to add later, if we want to require all properties to be listed in the binding. Suspect we might have lots of nodes that don't follow that now.
> 
> Probably not, unless this implementation (unlike the current one) generates defines for the extra properties. If there are any, they can't be being used.

Works like the old code. Thinking there might be a lot of unused stuff (or stuff that's undeclared because it's using some special path in the code). Maybe it's not that bad though...

File: scripts/dts/edtlib.py [hunk 3]
```diff
@@ -0,0 +1,1626 @@
...
+
+        paths = [path for path in self._binding_paths
+                 if os.path.basename(path) == filename]
+
+        if not paths:
+            _yaml_inc_error("'{}' not found".format(filename))
+
+        if len(paths) > 1:
+            _yaml_inc_error("multiple candidates for '{}' in !include: {}"
+                            .format(filename, ", ".join(paths)))
```
Comment (b0661) [inline scripts/dts/edtlib.py]: This prevents binding overload by application. An application (OOT) driver may need a special binding.
Using the first file in the list and reporting a warning would help in this case. An application can arrange to have it's bindings directory first in the list of directories searched.

Comment (ulfalizer) [reply scripts/dts/edtlib.py]: The old code also gives an error here, so it's not a behavior change at least.

Comment (b0661) [reply scripts/dts/edtlib.py]: Yes it is like in the old code. Just a remark. I patched it in my project to have an easy way to develop the OOT drivers.

File: scripts/dts/edtlib.py [hunk 4]
```diff
@@ -0,0 +1,1628 @@
...
+    def __init__(self, dts, bindings_dir):
+        """
+        EDT constructor. This is the top-level entry point to the library.
+
+        dts:
+          Path to device tree .dts file
+
+        bindings_dir:
+          Path to directory containing bindings, in YAML format. This directory
+          is recursively searched for .yaml files.
```
Comment (b0661) [inline scripts/dts/edtlib.py]: Should be a list of bindings directories to support applications that have their own bindings (additional to Zephyr).

File: scripts/dts/gen_defines.py [hunk 5]
```diff
@@ -329,6 +301,23 @@ def dev_instance_aliases(dev):
...
+    if not dev:
+        return
+
+    if not dev.regs:
+        err("missing 'reg' property in node pointed at by /chosen/{} ({!r})"
+            .format(prop_name, dev))
+
+    out_comment("/chosen/{} ({})".format(prop_name, dev.path))
+    out("{}_BASE_ADDRESS".format(prefix), hex(dev.regs[0].addr))
+    out("SRAM_SIZE", dev.regs[0].size//1024)
```
Comment (galak) [inline scripts/dts/gen_defines.py]: is this missing `prefix`?

Comment (ulfalizer) [reply scripts/dts/gen_defines.py]: Oops... fixed

**Review by erwango — DISMISSED** [review]
Can't build nucleo_f401re board (sample hello_world), due to missing symbol `CONFIG_RAM_SIZE`
It is defined in `arch/Kconfig`:
```
config SRAM_SIZE
	int "SRAM Size in kB"
	default $(dt_int_val,DT_SRAM_SIZE)
```
`DT_SRAM_SIZE` is not defined anymore

EDIT: Seems it has been replaced by:
```
#define DT_SRAM_SRAM_SIZE                           96
```
Cf fix in `scripts/dts/gen_defines.py`, line 318:
```
-    out("{}_SRAM_SIZE".format(prefix), dev.regs[0].size//1024)
+    out("{}_SIZE".format(prefix), dev.regs[0].size//1024)
```

**Comment by ulfalizer** [comment]
> Can't build nucleo_f401re board (sample hello_world), due to missing symbol `CONFIG_RAM_SIZE`
> It is defined in `arch/Kconfig`:
> 
> ```
> config SRAM_SIZE
> 	int "SRAM Size in kB"
> 	default $(dt_int_val,DT_SRAM_SIZE)
> ```
> 
> `DT_SRAM_SIZE` is not defined anymore

Fixed by the latest force push. Thought I could make a small change without diffing `generated_dts_board.conf` before and after, but I couldn't. :P

**Comment by erwango** [comment]
(Slack is broken to me so pasting my point here)
Would be nice to introduce nodes sorting at some step. 
From experience, this was useful in previous script when trying to debug and comparing 2 versions of generated output

**Comment by ulfalizer** [comment]
@erwango 
Since Slack is dead:

The order in the conf and header file matches the order in the .dts file now, so it should be consistent at least. Would be pretty easy to change though, if you have some case where that order is bad.

**Comment by erwango** [comment]
> The order in the conf and header file matches the order in the .dts file now, so it should be consistent at least. Would be pretty easy to change though, if you have some case where that order is bad.

This is not that I see, actually. For instance nucleo_f401re board, gpio nodes order:
`nucleo_f401re.dts_compiled`: A, B, C, D, E, H
`generated_dts_board_unfixed.h`: A, E, B, D, C, H (Changing at each build)

**Comment by ulfalizer** [comment]
@erwango 
Try this patch and see if it fixes it. `set()` iteration order can vary between runs in Python.

```diff
diff --git a/scripts/dts/gen_defines.py b/scripts/dts/gen_defines.py
index 720c297615..9d8c43a9c7 100755
--- a/scripts/dts/gen_defines.py
+++ b/scripts/dts/gen_defines.py
@@ -72,7 +72,7 @@ def main():
             active_compats.update(dev.compats)
 
     out_comment("Active compatibles (mentioned in DTS + binding found)")
-    for compat in active_compats:
+    for compat in sorted(active_compats):
         #define DT_COMPAT_<COMPAT> 1
         out("COMPAT_{}".format(str2ident(compat)), 1)

```

**Comment by ulfalizer** [comment]
@erwango 
What's your `python3 --version` btw? Dictionary iteration order didn't match insertion order before Python 3.6.

**Comment by erwango** [comment]
```
$ python3 --version
Python 3.5.2
```

Tried your proposal, but it didn't work (didn't help sorting devices). But as I understand, this aims at sorting compatibles and not devices, so this is logical.

**Comment by ulfalizer** [comment]
@erwango 
Yeah, the device-is-there flags use a `set()`, so they can move around between runs even on Python 3.6+, unless you sort them somehow.

The rest of it is consistent for me, so I'm guessing it's because you have Python 3.5.

**Review by erwango — APPROVED** [review]
